### ISSUE TITLE
On Ubuntu, neovide is not responsive when pinned and switching to it via meta+<number>

### ISSUE THREAD

**Issue by mrotaru** [neovide/neovide#942]
I'm used to using meta+[0-9] to switch between pinned programs - it works by default in both Windows and Linux.

I created an icon for neovide, and pinned it (using method described [here](https://askubuntu.com/a/1026532))- but when i switch to it via the meta shortcut, it's unresponsive - doesn't seem to be receiving any keyboard input. When I switch to it via alt-tab (or clicking the icon with the mouse), it's fine. 

Using latest neovide, compiled locally; Ubuntu 20.04.3 LTS; GNOME Shell 3.36.9

**Comment by rti** [comment]
Are you running X11 or Wayland? If it is X11, you can find some info about my findings in with a very similar issue [here](https://github.com/neovide/neovide/pull/931). 

Unfortunately I was unable to find an easy fix yet.

**Comment by rti** [comment]
I think I found a way to fix this issue for me. Really curious if this fixes the issue for you as well.

If you want to give it a try, you can build my branch `fix-gnome-x11-meta-mod-focus` from https://github.com/rti/neovide.

Actually, I had to change a bit of how winit handles key modifiers. More details can be found here https://github.com/rti/winit/commit/95f8e9294dbe8a7d382338465632b0f53814ed84. The `fix-gnome-x11-meta-mod-focus` branches on neovide and glutin are just there to pull in the winit change I made.

Would be very happy if you could give it a try.

**Comment by mrotaru** [comment]
@rti I'm on X11; can confirm the Gnome Tweaks mitigation you mention on #931 (making "Right Super (Meta)" the "Overview Shortcut) does work - however, as implied from the name, it has the side effect that the left meta can't be used to search for programs anymore.

I'll try your branch and let you know.

**Comment by mrotaru** [comment]
@rti it works !

Built your branch, turned off the mitigation in Tweaks, and was able to switch to it with meta+<number> - and still search for programs, using left meta.

**Comment by rti** [comment]
@mrotaru Perfect. Thanks for testing and reporting back.

**Comment by rti** [comment]
This _should_ be fixed in https://github.com/Kethku/winit/pull/1

**Comment by last-partizan** [comment]
@Kethku I'm trying to understand what is happening here, becouse i have issues with shift modifier being stuck, and i believe this is same issue, or at least very closely related.

I've got reliable way to reproduce it, but stumbled upon strange behaviour.

My reproduction is as follows:

```vim
map <S-Insert>		"*gP
cmap <S-Insert>		<C-R>*
imap <S-Insert>		<C-R>*
vnoremap <C-Insert> "*y
```

- Open `neovide`, press `Shift+Insert`, it should paste from clipboard.
- Release `Shift`, press only `Insert`, it pastes again. -- bug is here, shift modifier still present
- Press `Insert` again, it goes to INSERT mode, shift is unstuck.

I've added logging to neovide and winit, and SHIFT modifier is really stuck after first `Shift-Insert`

What bugs me - if i don't use those mappings, and just press `Shift-Insert`, it obviously does nothing. BUT, bug does not happen. After first `Shift-Insert` and releasing `Shift` - it is correctly released.

In bug situation it is stuck on glutin level, so i expected bug to be present even without my keymappings. So, my question is: does neovide send something to `winit`? or it's one-way communication?

I expected event flow roughly like this:

```
(macos|windows|x11|wayland) -> winit -> neovide <-> neovim
```

And, bug happens for me only in x11 mode, wayland works just fine, so i'm sure there is something wrong with winit/x11.

Here is some logs, from neovide and printlns from winit:
<details>
<summary>Without mappings, normal case</summary>

```
 > keyperss or keyrelease: ShiftRight, (empty)
 if Some(modifier)
   > modifier: Shift, state: Pressed
    > self.keys: {}, mod: Shift
     > key_press: 62, modifier set: Shift
   > new modifiers: SHIFT
INFO [neovide::window::keyboard_manager] mods: SHIFT
 > keyperss or keyrelease: Insert, SHIFT
INFO [neovide::window::keyboard_manager] <S-Insert> shift: true, ctrl: false, alt: false, logo: false
 > keyperss or keyrelease: Insert, SHIFT
 > keyperss or keyrelease: ShiftRight, SHIFT
 if Some(modifier)
   > modifier: Shift, state: Released
    > self.keys: {62: Shift}, mod: Shift
     > key_release: 62
     > modifier removed: Shift
   > new modifiers: (empty)
INFO [neovide::window::keyboard_manager] mods: (empty)
```

</details>

<details>
<summary>With mappings, bug case</summary>

```
 > keyperss or keyrelease: ShiftRight, (empty)
 if Some(modifier)
   > modifier: Shift, state: Pressed
    > self.keys: {}, mod: Shift
     > key_press: 62, modifier set: Shift
   > new modifiers: SHIFT
INFO [neovide::window::keyboard_manager] mods: SHIFT
 > keyperss or keyrelease: Insert, SHIFT
INFO [neovide::window::keyboard_manager] <S-Insert> shift: true, ctrl: false, alt: false, logo: false

# Here is strange part: just after sending S-Insert, keyboard_manager received several ModifiersChanged #

INFO [neovide::window::keyboard_manager] mods: (empty)
INFO [neovide::window::keyboard_manager] mods: (empty)
INFO [neovide::window::keyboard_manager] mods: SHIFT
INFO [neovide::window::keyboard_manager] mods: (empty)
INFO [neovide::window::keyboard_manager] mods: SHIFT
 > keyperss or keyrelease: Insert, SHIFT
 > keyperss or keyrelease: ShiftRight, SHIFT
 if Some(modifier)
   > modifier: Shift, state: Released
    > self.keys: {}, mod: Shift
     > key_release: 62
   > new modifiers: SHIFT
INFO [neovide::window::keyboard_manager] mods: (empty)
```

</details>

**Comment by last-partizan** [comment]
I found what's happening here, comminication is indeed one-way from x11 to winit, but somehow 

```vim
map <S-Insert>		"*gP
```

When executing this, window loses focus, and regains it several times, and during this glutin sends ModifiersChanged events.

**Comment by Kethku** [comment]
Wtf?? That's incredibly strange. What are the chances this is due to the new keyboard input handling in upstream?

**Comment by Kethku** [comment]
I guess pretty low because that's an action from neovim. Can you embed the log? I'm curious what events we are getting from nvim

**Comment by last-partizan** [comment]
Here's log, look at FocusLost/Gained events.

This is log from x11.
[neovide_rCURRENT.log](https://github.com/neovide/neovide/files/7927993/neovide_rCURRENT.log)

This is log from wayland:
[neovide_rCURRENT.log](https://github.com/neovide/neovide/files/7928018/neovide_rCURRENT.log)

Wayland too has FocusLost/Gained events, but it glutin handles it correctly. And i'm almost fixed it under x11.

**Comment by last-partizan** [comment]
Even if those strange focusing issues are from neovim, or whatever, it could be completely fixed at glutin level.

**Comment by last-partizan** [comment]
x11 handling is absolute mess in winit.

https://github.com/neovide/winit/commit/2fe2f620a3c3863806e381fc7204da3ab179feb7

Right now i've got it working, mostly by removing WTF parts and by adding lots of debug printfs, it's already working as expected, and i plan to test it for a few days, and then create separate branch in neovide where we could test it together.

**Comment by Kethku** [comment]
Awesome work. I'm pretty sure we will have to contribute w/e changes upstream (probably the good thing to do anyways) so its going to take some doing, but this is a great first step.

**Comment by last-partizan** [comment]
There is good news and the bad news.

I have following mapping:

```vim
set pastetoggle=<F4>
```

The bad news: when i press `Alt+F4`, neovide instead of closing goes into paste mode. When i press any non-modifier key, it closes.
The good news: it reproduces on main branch, so maybe it's even unrelated not only to my changes, but to neovide in general. But it was working just fine yesterday.

**Comment by last-partizan** [comment]
Yeah, nevermind, it has nothing to do with neovide, this is Gnome fucking around.

**Comment by last-partizan** [comment]
@mrotaru please, check out this branch: https://github.com/neovide/neovide/tree/fix-stuck-shift

It should fix this issue.

**Comment by fredizzimo** [comment]
I think the new winit version merged in https://github.com/neovide/neovide/pull/1789 fixes the issue. Could you test with Neovide master?

**Comment by fredizzimo** [comment]
The potential fix is now included in Neovide 0.11.0, could you test with that?

**Comment by fredizzimo** [comment]
Closing, since no new reports after the release
